### ISSUE TITLE
CSSOM View document.scrollingElement

### ISSUE THREAD

**Issue by zcorpan** [w3ctag/spec-reviews#51]
The Blink team would like a TAG review of document.scrollingElement before LGTM-ing shipping it.
https://groups.google.com/a/chromium.org/d/msg/blink-dev/mJ5SlCj3CHc/hkQuqROnwl4J
https://groups.google.com/a/chromium.org/d/msg/blink-dev/mJ5SlCj3CHc/SnJ5rNxCiOUJ

Spec: http://dev.w3.org/csswg/cssom-view/#dom-document-scrollingelement

Background/rationale
https://github.com/operasoftware/devopera/issues/242
https://lists.w3.org/Archives/Public/www-style/2015Apr/0108.html

Tracking bugs
https://code.google.com/p/chromium/issues/detail?id=475531
https://bugzilla.mozilla.org/show_bug.cgi?id=1153322
https://bugs.webkit.org/show_bug.cgi?id=143609

**Comment by foolip** [comment]
The next branch point is on May 15th, 2015, so it would be very good to have any feedback well before that in case we want to ship with the next milestone.

**Comment by domenic** [comment]
I'm personally unclear why this is the right path. Given the existence of e.g. https://github.com/mathiasbynens/document.scrollingElement it seems like we are essentially baking a jQuery-style browser normalizer ... into the browser. Which means we'll have to support it going forward forever.

Why does this drop-in replacement need to be provided by the browser? It's marginally easier for authors, but is it worth the long-term cost of adding cruft, in the future where document.scrollingElement just becomes a synonym for document.documentElement? You have to consider the cognitive burden for future authors who read code and see document.scrollingElement, then have to go look up what that does, instead of just seeing document.documentElement and knowing immediately what it does. (And yes, it's not just a synonym in quirks mode, but I don't think adding a new API for quirks-mode users is really that interesting).

I also don't understand how we expect this to work out. Is the envisioned sequence something like this?

- spec says document.scrollingElement === document.documentElement
- Blink implements document.scrollingElement === document.body; Gecko implements document.scrollingElement === document.documentElement
- Hopefully nobody writes code that assumes one or the other (e.g. uses document.scrollingElement.firstChild or document.scrollingElement.addEventListener(...)), or uses UA-sniffing such that different code paths within the same app depend on the different behaviors for document.scrollingElement
- We try switching to standards-compliant behavior?

I am worried about the "hopefully" here, as it just seems to set us up for repeating the cycle. Likely I am missing something.

**Comment by RByers** [comment]
I agree there is some cause for concern here.  I believe your argument is only compelling when your ignore quirks mode.  Even in the future when Blink/WebKit are fixed, libraries still need a bit of logic to determine which element is the scrolling element.

Trying to get sites to update to support the spec compliant behavior without a UA check it'be become clear to me that scrollingElement is a big help in practice.  Facebook made use of such an API internally, was unwilling to do the (somewhat intrusive iframe-measurement) hack and so had to make a fairly substantial change to their codebase to follow separate get ('||' the two values) and set paths (set on both elements).  Closure is even worse - their public API has 'goog.dom.getDocumentScrollElement'.  We course of course could argue that they (and anyone else like them) should use an iframe measurement hack (like the above polyfill - which by the way still has issues), or they should deprecate this public API.  But that would take way more debate/effort than at least I'm willing to stomach.

If we think the value of document.scrollingElement isn't high enough, then that's OK with me.  I'll give up trying to ever fix this wart in blink and be resigned to either the status quo (continuing the LONG tradition of violating the CSS spec here) or trying to convince the spec and other browsers to change  to bake WebKits bug into their design (thereby making body-scrollers confusingly unavailable to JavaScript).

So to me the choice of long term future seems to be between:
1) All browsers eventually behave the same and we have a somewhat questionable tiny additional API
2) All browsers have irrational behavior where body doesn't truly behave like an element, and documentElement controls scrolling via some APIs (CSS overflow) and not others (scrollTop)
3) Status quo - confusing differences between browsers

I'd be happy to be proven otherwise.  There have been several failed attempts over the years to fix this wart, but adding a new API to ease the transition is the one thing that hasn't been tried.  Perhaps I'm just crazy for trying again to solve this (after IE told me they were forced to copy our bug in EdgeHTML).

**Comment by domenic** [comment]
> I believe your argument is only compelling when your ignore quirks mode. Even in the future when Blink/WebKit are fixed, libraries still need a bit of logic to determine which element is the scrolling element.

The scenario here is a library that intends to work in quirks mode as well as standards mode? Is that a widespread need? If so that would indeed be a good argument.

The rest of your post explains why you think this will work but doesn't really address my concern for why I think it could not work. In other words, I think you are missing long-term futures like

4) All browsers have irrational behavior where body doesn't truly behave like an element, and documentElement controls scrolling via some APIs (CSS overflow) and not others (scrollTop), and browsers also have a scrollingElement alias for body

or

5) Status quo - confusing differences between browsers - including scrollingElement now also being a new difference

**Comment by foolip** [comment]
> IE told me they were forced to copy our bug in EdgeHTML

Ouch, so this has already happened. Did they tell you which kinds of sites were fixed by this? Is it just sites that sniff for WebKit, or is it more intricate?

**Comment by RByers** [comment]
> seems like we are essentially baking a jQuery-style browser normalizer ... into the browser. Which means we'll have to support it going forward forever.

Isn't this how browser APIs evolve to some extent?  As frameworks figure out better API surface area to reduce developer confusion and interop problems, we _should_ consider adopting it into the core platform for universal availability?

> The scenario here is a library that intends to work in quirks mode as well as standards mode? Is that a widespread need? If so that would indeed be a good argument. Are libraries currently using tricks like those on display in the polyfill before they use scrollTop, in order to cater to quirks mode?

I don't know how common this is.  Many of the solutions I've seen for this "how do I identify the scrolling element" question involve checking for quirks mode.  Eg. see Closure's code here: https://github.com/google/closure-library/blob/32365aba43acb36c5d693256ef5d4dbe3bddddfe/closure/goog/dom/dom.js#L632

From what I can see, jQuery deals with this issue by creating unified scrollTop/scrollLeft wrapper APIs which use the window api when appropriate (eg. window.pageXOffset) rather than attempt to use body/documentElement at all.  Eg: https://github.com/jquery/jquery/blob/74ae5444832b2fb966768a97281d2ad8c088bc58/src/offset.js#L170.

> I think you are missing long-term futures like ...

Yes, I agree those outcomes are a risk and I don't have any concrete data / arguments to counter it.  My instinct is that the cost (complexity/confusion) of a precisely defined but extraneous scrollingElement API is relatively low compared to the cost of the other options.  Also that it's worth taking some small breakage to fix this issue, and therefore the benefit (over a suitably long time horizon) outweighs the risk that we'll never fix it.  

But I don't see any obvious way around this being a judgement call.  It sounds like Safari agrees with your judgement ([same sort of argument](https://bugs.webkit.org/show_bug.cgi?id=143609)), while Mozilla, IE and the spec editor's judgement is that it's worth it.  I'm happy to defer to your judgement if TAG should be the one to make the judgement call here.

**Comment by RByers** [comment]
> Ouch, so this has already happened. Did they tell you which kinds of sites were fixed by this? Is it just sites that sniff for WebKit, or is it more intricate?

From what I've heard, it's only due to UA sniffing.  EdgeHTML includes 'Webkit' in it's UA, so in general plans to copy WebKit bugs where it matters for interop and we don't have a clear plan to solve them.  My hope was that we could still solve this before it gets worse - if I can make the change stick in blink then EdgeHTML and WebKit would hopefully find it relatively easy to follow suit.

**Comment by mathiasbynens** [comment]
Some more background on why moving this from user code to the browser might be a good thing: feature-detecting the scrolling element (especially if you need to get the edge cases right, which is what libraries usually aim for) is costly. One way is to actually scroll the document:

```js
var $scrollElement = (function() {
  var $html = $(document.documentElement);
  var $body = $(document.body);
  var bodyScrollTop;
  if ($html.scrollTop()) {
    return $html;
  } else {
    bodyScrollTop = $body.scrollTop();
    // Scrolling the body doesn’t do anything.
    if ($body.scrollTop(bodyScrollTop + 1).scrollTop() == bodyScrollTop) {
      return $html;
    } else {
      // We actually scrolled, so undo it.
      return $body.scrollTop(bodyScrollTop);
    }
  }
}());
```

The terribleness might explain why so many (e.g. Google, Facebook) opt to UA sniff instead, which is more performant but leads to compat problems (@rbyers knows all about them).

**Comment by mikesherov** [comment]
This originally surfaced as a bug in jQuery/jQuery UI. IMO, scrollingElement is just another hack around the issue. The canonical, non-UA sniff, AFAIK, is to do `$html.scrollTop() || $body.scrollTop()`. The *problem* was that lots of people were using `$html.scrollTop() + $body.scrollTop()`, which is problematic when the body or the html element have `overflow: scroll;` applied.

**Comment by RByers** [comment]
Mike, do you have a suggestion for how we get frameworks like closure to update without adding an API like scrollingElement?  I'm hearing a bunch of objection to scrollingElement, but I haven't yet heard anyone suggest a realistic path for actually fixing the problem here.  As with most compatibility vs. rationality debates the problem is over constrained.  If we continue to hold out for a perfect ideal future, we'll remain at our currently horribly confusing state IMHO.

**Comment by mikesherov** [comment]
The bug was inherited from Webkit, who has since ripped the bandaid off.

If the future is that scrollingElement is just another stop gap, and already has a polyfill, we point the libs to the polyfill until they can wrap their heads around just always using document.documentElement.

jQuery has worked around this in UI for a while. In fact, we're eagerly awaiting the bandaid to be ripped off simply so we can remove our workarounds.

Perhaps I'm misunderstanding, but the UA sniffs and the polyfills are all because Blink hasn't bit the bullet yet.

Shipping scrollingElement won't alleviate need for polyfill of old blink, so nothing is gained by introduction of the new API?

**Comment by RByers** [comment]
Wait, what do you mean WebKit has ripped the bandaid off?  There has been an open bug there for 9 years ([this](https://bugs.webkit.org/show_bug.cgi?id=106133) and previously [this](https://bugs.webkit.org/show_bug.cgi?id=9248)) which they've tried and failed to fix (for similar reasons as us).  If they successfully changed then I have no doubt we could too.  But instead the opposite is happening - IE has switched to using 'WebKit' in their user agent and due to compat issues felt they had to also copy this bug.

Shipping scrollingElement will alleviate the need for an awkward polyfill - sites are free to fall back on the simple UA check they're already doing today when scrollingElement isn't defined.  scrollingElement gives us a simple drop-in upgrade path for such sites.

**Comment by mikesherov** [comment]
I missed that Webkit had to revert. I was eagerly watching it land initially. My apologies for the mistake.

It still just seems like a weird wart to me to have to add API surface simply for feature detecting the presence of a bug.

In either case, I think having a canonical polyfill is good enough, but for me, anything we can do to move the web forward here works.

I'm tired of writing `$('html, body').scrollTop(0);` to get back to the top of the page in a cross browser way.

**Comment by RByers** [comment]
Domenic, please take a look at the current (now believed to be correct) implementation of the scrollingElement polyfill: https://github.com/mathiasbynens/document.scrollingElement/blob/master/scrollingelement.js.

Hopefully the complexity / subtlety here helps justify why it'll be a tough sell to get frameworks/apps to fix their bug by replacing a 1-line UA check with this 100-line polyfill.

**Comment by mikesherov** [comment]
To be fair, more than a few of those lines are to support IE4 and IE7. Also a few of the lines ensure its a property rather than a function that can be called. Also, if I'm reading it right, it always returns document.documentElement is non-quirks mode? 

Does FB or Closure support non-quirks mode or any of those effected by this bug actually support the browsers that take up a good portion of the logic?

**Comment by mathiasbynens** [comment]
> Also, if I'm reading it right, it always returns document.documentElement is non-quirks mode?

Not in buggy WebKit/Blink browsers.

**Comment by RByers** [comment]
> Does FB or Closure support non-quirks mode or any of those effected by this bug actually support the browsers that take up a good portion of the logic?

Closure at least [supports quirks mode](https://github.com/google/closure-library/blob/32365aba43acb36c5d693256ef5d4dbe3bddddfe/closure/goog/dom/dom.js#L632) - yes.  Can't say about facebook.

**Comment by domenic** [comment]
@RByers I really just don't think it makes sense to ship that kind of polyfill with a browser. You've found one library (a Google library!) which directly exposes the scrolling element. Because they've purposefully added such complexity, it seems totally reasonable for them to eat it and ship the few lines necessary for non-IE4/IE7 support. Others can use much simpler tricks that have been cited in this thread. That said, I'm not the one on the ground here, and if you really think this'll make a difference (although IMO Closure really should not count), then maybe I'm just wrong.

Anyway, I think I've expressed my opinion enough on this thread, especially since I'm just ex-TAG. We'll let others chime in.

**Comment by zcorpan** [comment]
It might be worthwhile to point out that a native implementation does not have the same complexity as the polyfill. The "scrolling element" is a concept that the browser needs anyway to implement `scrollTop` et al, so it's just a matter of exposing that concept with an IDL attribute.

**Comment by mathiasbynens** [comment]
> The “scrolling element” is a concept that the browser needs anyway to implement `scrollTop` et al, so it’s just a matter of exposing that concept with an IDL attribute.

This makes it sound like a no-brainer, given [the Extensible Web Manifesto](https://extensiblewebmanifesto.org/).

**Comment by mikesherov** [comment]
So, just so that I'm clear, can you show a code example of how FB or jQuery or Closure uses this property, including cases where it isn't present?

**Comment by zcorpan** [comment]
With polyfill:
```html
<!-- in <head> if necessary, otherwise before </body> -->
<script src="scrollingelement.js"></script>
```
```js
document.scrollingElement.scrollTop = 100;
```

without polyfill, fall back to UA check
```js
function getScrollingElement() {
    if ("scrollingElement" in document) {
        return document.scrollingElement;
    }
    return /* whatever UA check they use now */;
}
```
```js
getScrollingElement().scrollTop = 100;
```

**Comment by RByers** [comment]
I've talked with the Closure developer for this today.  He is [unwilling to use the iframe measurement hack as it's too expensive](https://github.com/google/closure-library/issues/455).  I.e. scrollingElement or something like it appears to be necessary to get most Google properties updated.

**Comment by mikesherov** [comment]
@RByers Again, I'm a layman, so please just take my comments as feedback. I've also been trying to solve this problem for a while on behalf of jQuery, and this is as close as we've ever gotten to a solution....

what about a boolean property rather than an additional reference to document.documentElement? I think @domenic and my (layman) main objection is about yet another property referring to the same underlying element. Perhaps `document.documentElement.isScrollingElement`?

This would look like:

```
function getScrollingElement() {
    if (document.documentElement.isScrollingElement) {
        return document.documentElement;
    }
    return /* whatever UA check they use now */;
}
```

You'd also have the benefit of querying an element to see if it is the scrolling element without having to compare it to another element:

```
function isScrollingElement(element) {
  if ("isScrollingElement" in element) {
    return element.isScrollingElement;
  }
  return /*some UA check */;
}
```

**Comment by RByers** [comment]
Sure, that would be fine with me as well.  You could argue that adding to 'Document' is a smaller change than adding to 'Element', but then we could have 'boolean Document.bodyScrollsDocument' or some such.  I'm happy to take any suggested tweaks like that to the CSS WG and other vendors.

**Comment by mathiasbynens** [comment]
@mikesherov 

> what about a boolean property rather than an additional reference to document.documentElement?

It’s not just another reference to a fixed value. `document.scrollingElement` would only be a reference to `document.documentElement` in standards mode and if the browser implements the scrolling element as per the spec. I.e. if WebKit were to implement `document.scrollingElement` before fixing their bug, `document.scrollingElement` would not point to `document.documentElement` even in standards mode.

> […] This would look like:
>
> ```js
> function getScrollingElement() {
>   if (document.documentElement.isScrollingElement) {
>     return document.documentElement;
>   }
>   return /* whatever UA check they use now */;
> }
> ```

No, it would then have to look like:

```js
function getScrollingElement() {
  if (document.documentElement.isScrollingElement) {
    return document.documentElement;
  }
  if (document.body.isScrollingElement) {
    return document.body;
  }
  return /* whatever UA check they use now */;
}
```

**Comment by mnot** [comment]
Discussed at SF F2F; Alex will draft a review.

**Comment by mikesherov** [comment]
@mathiasbynens sure, that makes sense. Thanks for walking me through it. I'll state my case one last time, and then be done with it :-)

I suppose that my main objection is simply that I'd prefer to just feature detect that this bug is fixed, and go on my merry way of behaving the way the spec is written, rather than augmenting the spec to including an additional reference (even though it changes based on quirks mode or not) to either `document.documentElement` or `document.body`. Seems like less to have to change in the future, and less confusing for future web authors who never use quirks mode, and never developed during a time in which this bug existed, as to why there are 2 properties referencing document.documentElement.

If I have a simple boolean property, I can just replace `/* whatever UA check they use now */;` with `document.behavesCorrectly || /* whatever UA check they use now */;`

In closure, the change would be:

```
goog.dom.getDocumentScrollElement_ = function(doc) {
  // WebKit needs body.scrollLeft in both quirks mode and strict mode. We also
  // default to the documentElement if the document does not have a body (e.g.
  // a SVG document).
  if (!goog.userAgent.WEBKIT && goog.dom.isCss1CompatMode_(doc)) {
    return doc.documentElement;
  }
  return doc.body || doc.documentElement;
};
```

to:

```
goog.dom.getDocumentScrollElement_ = function(doc) {
  // WebKit needs body.scrollLeft in both quirks mode and strict mode. We also
  // default to the documentElement if the document does not have a body (e.g.
  // a SVG document).
  if ((!goog.userAgent.WEBKIT || document.behavesCorrectly) && goog.dom.isCss1CompatMode_(doc)) {
    return doc.documentElement;
  }
  return doc.body || doc.documentElement;
};
```

as opposed to:

```
goog.dom.getDocumentScrollElement_ = function(doc) {
  if ("scrollingElement" in document) {
    return document.scrollingElement;
  }
  // WebKit needs body.scrollLeft in both quirks mode and strict mode. We also
  // default to the documentElement if the document does not have a body (e.g.
  // a SVG document).
  if (!goog.userAgent.WEBKIT && goog.dom.isCss1CompatMode_(doc)) {
    return doc.documentElement;
  }
  return doc.body || doc.documentElement;
};
```

In the end, I commend @RByers for pushing this towards actually getting addressed, and I'm happy with whichever solution ends up happening here. Just wanted to make sure the case was made for alternate approaches.

**Comment by zcorpan** [comment]
So in the future, assuming the bug is successfully fixed, and you don't have to worry about current-era browsers, but still worry about quirks mode (like Closure does), with `scrollingElement` it would be like this:

```js
goog.dom.getDocumentScrollElement_ = function(doc) {
    return doc.scrollingElement;
}
```
(or scrap the method altogether and use `scrollingElement` directly)

with `behavesCorrectly` it would be (assuming it returns `false` in quirks mode?)

```js
goog.dom.getDocumentScrollElement_ = function(doc) {
    if (doc.behavesCorrectly) {
        return doc.documentElement;
    }
    return doc.body || doc.documentElement;
}
```
However this would not be technically correct (the current Closure script isn't, either), because `document.body` can be a `frameset` element, or the `body` can be independently scrollable (in which case, `scrollTop` should scroll the element, not the viewport, per spec), or if `document.body` is null it would give you the `html` element but that is not the "scrolling element" in quirks mode, and so on. Possibly being technically correct is a non-goal for Closure; it was a goal for the polyfill. With `scrollingElement` you get correct behavior "for free".

**Comment by dperini** [comment]
Shouldn't the `scrollingElement` property be part of the `viewport` instead of the `document` ?
Thus `document.defaultView.scrollingElement` or simply `window.scrollingElement`.

**Comment by zcorpan** [comment]
* It's supposed to be a drop-in replacement for `document.body` or `document.documentElement`, which are on `document`.
* Adding a property on the global is higher risk than adding a property on `document`.

**Comment by dperini** [comment]
@zcorpan thank you for the relevant point of view, I understand your reasoning behind both of them.
However some of the late additions to the global do not follow that exact reasoning.
For example API like `window.requestAnimationFrame()` and `window.cancelAnimationFrame()`, or speaking of recent properties additions `window.localStorage` and `window.sessionStorage` which are the new `document.cookie` alternatives/replacements.
For something older in time, see the `window.location` property which is a duplicate of the `document.location` property.
There are surely other examples of this.

**Comment by phistuck** [comment]
@dperini - do you have an example of a property of `window` that refers to an element (not `document`)?
`for (x in window) { try { if ("nodeType" in window[x]) console.log(x) } catch (e) {}}`
Only outputs -
`document`
For me.

**Comment by dperini** [comment]
@phistuck,
no I don’t have such an example, but I don’t think that’s the way to measure the correctness of what is being implemented (wrongly in my humble opinion).

Other than that, my proposal was to have a `getScrollingElement` polyfill or a new method on the global space and not a new `scrollingElement` property on the `document`. There may already be enough aliases of the `documentElement/body` on the `document` node if we look at "DOM 1 Core" and the "DOM 2 Traversal".

This proposal does fit with your reasoning on the above code snippet and is future proof for when Webkit would fix the bug about Quirks/Standards modes behavior related to which element scrolls the `viewport`. That behavior have been in the specs for the last decade, the bug was spotted a long time ago but it was never fixed in Webkit. Now that bug was "forcedly" inherited also by Internet Explorer and Opera. It is the right time to fix it in the right way, instead of forcing everybody to make the wrong choice and the specs rewritten.

Have you noticed that everything "scrolling" related (properties and methods) are already bound to the `viewport` ? (see: scrollX, srcollY, scroll(), scrollBy(), scrollTo() and more).

Have you noticed that there is nothing "scrolling" related on the `document` node ?

**Comment by phistuck** [comment]
@dperini - I am not sure mixing DOM concepts (DOM elements) and `window` (which is a `global`) is the right way to go but I defer to others for the rationale behind not adding such properties to `window` so far (I guess there is a rationale).

By the way, the methods you mentioned look like legacy methods (introduced by a single browser in the first browser wars and copied by others for compatibility or competitiveness) to me and not a result of consensus or deep thinking, really.
And, yes, I know they are specified as a part of the CSSOM View Module, but that does not mean a lot (it may merely document the APIs that are already prevalent and you cannot get rid of them, so they better be specified than otherwise).

**Comment by dperini** [comment]
@phistuck,
@phistuck,
just to say it clearer ... I am against adding a `scrollingElement` property to the `document`, it is already clear in the specs in which cases the scrolling element will be the `body` element (only in Quirks Mode), in all other cases the scrolling element should be the `documentElement`.
In any case, if that's deemed necessary to ease future detection I suggest adding a new method to the `viewport` (the global `window`).

Keep in mind that adding this new property or method doesn't solve anything for those that felt they should trust UA sniffing (the primary concern), they will be forced to change their code anyway if the bug is going to be fixed one day. The other way around is changing the specs and copy the bug also on Firefox and other already spec compliant engine. The latter, introducing far more changes in code written in the last decade.

My 2 cents :)

**Comment by dperini** [comment]
@phistuck,
just for fun, but following the reasoning in the above code snippet, you can see that on the `document`, excluding the Traversal properties, there is only one property pointing to an `Element` whose `parentNode` is a `document` node (nodeType == 9):

    for (x in document) {
        try {
            if (!/Child$/.test(x) && document[x].nodeType == 1) {
                if (document[x].parentNode.nodeType == 9) console.log(x)
            }
        } catch (e) {}
    }

which yields only one property:

    documentElement

adding `scrollingElement` property to `document` will change the results to yield two properties:

    documentElement
    scrollingElement

**Comment by dperini** [comment]
Just noticed that in ChromeCanary, when enabling the "Experimental Web Platform Features" in a frameset document, the `document.scrollingElement` may be a reference to the frameset element.
Also, for a newly created document the same property value will be `null`. Not sure it is expected.

**Comment by dperini** [comment]
Also noticed that Safari is missing some of the "DOM 2 Traversal" properties/attributes on the `document`. Both Firefox and Chrome have `firstElementChild` and `lastElementChild` properties on the `document` node.

**Comment by phistuck** [comment]
@dperini - why do you only look for nodes that have `document` as their parent node? It is a meaningless constraint (to me, anyway).
`document.body`, `document.head`, `document.images[0]` and so and so (yes, traversal as well) return nodes. Nothing on `window` does, except `document` which is the only entry point to the DOM.

Anyway, I am just stating the current state of affairs. It does not mean that there will never be relations from `window` to elements. If you still do not see my point or disagree, there is not much I can do to explain it and this discussion is pretty pointless, since I am certainly not an expert (but the TAG team is), so I defer to the experts.

**Comment by RByers** [comment]
That is expected.  document.body sometimes returns a frameset](http://www.w3.org/TR/html5/dom.html#the-body-element-0) too, and [document.documentElement should be null in the empty document case](https://dom.spec.whatwg.org/#document-element).

**Comment by dperini** [comment]
@phistuck,
I didn't want to sound disrespectful, I am not an expert either, these were just my observations.
As there is only one property on `window` to drill-down to the DOM tree, there is only one property on `document` referring to an element having the `parentNode` to walk-up directly to the DOM tree host object, the `document` itself. But, I repeat, it was "just for fun" I am not trying to draw conclusions about those coincidences.

**Comment by dperini** [comment]
@RByers thank you for the answer, the clarification and the links to the relevant specifications.

**Comment by zcorpan** [comment]
@RByers
> That is expected.

The frameset thing is a bug, the CSSOM View spec for `scrollingElement` does never return a frameset element. It specifically calls this out in a note. :-)

@dperini It seems to me that your line of reasoning is that the `window` is equivalent to "the viewport", but I think that is not the case. There is no object actually representing the viewport. So for some APIs it's the window, for some it's the document, for some it's the root element, for some it's the body element.

**Comment by dperini** [comment]
@zcorpan,
> It seems to me that your line of reasoning is that the window is equivalent to "the viewport"

from what I understand this is almost exact, a more elaborate wording to describe that could be:

in a desktop browser, showing an non styled document, the `window` is equivalent to the `document.defaultView` which is the default document viewport ("porthole").

In mobile browsers having vendor supplied UA stylesheet, or in desktop browsers having user supplied stylesheet, the supplied style may affect and define the number of currently available viewports ("portholes"). This may be achieved through the `overflow`, `width`, and `height` CSS properties and their values.

To summarize ... there may be more than one viewport for an HTML document, though there is only one `defaultView`.

**Comment by dperini** [comment]
@zcorpan,
basically each DOM Element could be made to represent a `viewport` through the use of CSS `overflow` property combined with the `width` and `height` properties.

**Comment by slightlyoff** [comment]
Hey all,

Thanks for the thoughtful discussion here.

Per most transitional periods for features, we might see pain for a long time as a result of any change. That means we aren't going to take a side about what Blink should do to manage a transition (if any).  

The TAG's job is to make sure that _when_ we arrive at a different world, it's the one that's actually better. A functional DOM that doesn't require wrapping ad infinitum is a goal we must aim for.

Lastly, multiple viewports aren't A Thing (TM) in the platform today. Many APIs we expose simply won't do the right thing in a multi-viewport world. They're the vestigial tail of a once hoped-for future. They do not weigh on the discussion at hand.

The current situation is bad/broken and that fixes which exacerbate the `+` issue won't fly. We also view the `<iframe>` solution as a non-starter. As a result, I'm going to write up a review that captures as much of the discussion here as seems reasonable with the punch-line being that the Blink proposal looks good.

The larger concern I've got is that none of this reads on the root issue: how did `scrollingElement` get decided? How can a programmer set it themselves? I.e., why is this magical? That'll be the meat of the discussion in the writeup.

Regards

**Comment by dperini** [comment]
@slightlyoff,
handling multiple viewports has never been the objective here, it was only a discussion follow-up.
I believe that https://github.com/dperini/scrollToAnchorRef project better explains my point on why I absolutely needed to determine the `scrollingElement` of the current `defaultView`in a cross-browser way.

**Comment by RByers** [comment]
@zcorpan 

> The frameset thing is a bug, the CSSOM View spec for scrollingElement does never return a frameset element. It specifically calls this out in a note. :-)

Ah, thanks.  I see now that I recently introduced [a bug](http://crbug.com/484720) here (where document.body.scrollTop would scroll the viewport even when body is a frameset).  I will get that fixed before it hits beta.

**Comment by travisleithead** [comment]
Point of information: this has recently been implemented in "Microsoft Edge" (Project Spartan), as per the agreement in: https://groups.google.com/a/chromium.org/forum/#!searchin/blink-dev/scrollingelement/blink-dev/mJ5SlCj3CHc/PNSSYY5O_IQJ

It solves the problem of working around Blink's viewport bug (which Edge copied for interop), and also helps with the Quirks Mode use case. So, seems OK to me :-)

**Comment by torgo** [comment]
Discussed at Boston f2f.

**Comment by torgo** [comment]
@slightlyoff to write up some feedback.

**Comment by slightlyoff** [comment]
We didn't get to a formal review doc on this, but provided feedback in comment here: https://github.com/w3ctag/spec-reviews/issues/51#issuecomment-98203476

Implementers converged on this behavior and the open item about setting `.scrollingElement` seems like something for the Houdini and CSS folks to wrestle with another day.

Closing, given that events converged on the TAG's preferred outcome.
